Implement @depositBits and @extractBits

bnuuydev opened the pull request:
This PR takes #18680, updates the changes to master, and fixes the outstanding issues which prevented it from being merged.

There currently are outstanding issues related to the COFF linker which are blocking progress in the x86 backend:

- #19498 causes causing calls to the compiler-rt functions for u128 to crash the compiler when using the x86 backend to target Windows.
- All calls to the compiler-rt functions for pext and pdep appear to invoke a bug in the COFF linker, causing these symbols to be linked from a non-existent DLL called `.dll`. 

I have been told that moving on and just skipping the failing tests is the appropriate course of action here.

The comptime implementation calls into `big.int.depositBits/extractBits`, which themselves call the corresponding builtins. This allows for significantly better performance at comptime, but obviously requires `zig1.wasm` to be updated. I understand that a core member will be required to redo this commit for me before this can be merged?

The C backend implementation for integer sizes above `u64` is blocked by #19991. I am subscribed to this issue, and once fixed, I have a work-in-progress implementation ready to be completed and included in a follow-up PR. Up-to `u64`, the C backend implementation works fine, and is sufficent for bootstrapping the compiler.

I've included a number of behaviour tests to ensure that the implementations are correct.

Many thanks to @Rexicon226, @jacobly0 and others for your extensive help :)

Closes #14995

bnuuydev commented:
There's currently a copy-paste in compiler-rt waiting for the required functions to be moved to `std/Target.zig`. Once that change is made, I'll fix that commit up.

bnuuydev commented:
There is scope for a follow-up PR to optimise the codegen for `u128` and above to compile down to multiple invocations of `pdep` and `pext`. I intend to tackle that at a later date :)

jacobly0 commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                            .src_constraints = .{ .{ .int = .byte }, .{ .int = .byte }, .any },
+                            .patterns = &.{.{ .src = .{
+                                .{ .to_reg = .dil },
+                                if (air_tag == .extract_bits) .to_gpr else .{ .to_reg = .sil },
```
I think you'll need two cases here, one which matches against `.mem` and has a temp for `.{ .type = .u32, .kind = .{ .param_gpr = .{ .cc = .ccc, .index = 1 } },`  and always does `.{ ._, ._, .movzx, .tmp0d, .src1b, ._, ._ }`, and another which matches against `.{ .to_param_gpr = .{ .cc = .ccc, .index = 1 } }` and conditionally does `.{ ._, ._, .movzx, .src0d, .src0b, ._, ._ },`.  This avoids allocating a callee saved register with `.{ .to_reg = ... }` for a value that isn't even live across any calls

bnuuydev replied:
I think I could do pretty much the same as you describe for the `movzx`es with the native instructions too.

jacobly0 commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                    },
+                    else => unreachable,
+                }) catch |err| switch (err) {
+                    error.SelectFailed => return cg.fail("failed to select {s} {} {}", .{
+                        @tagName(tag),
+                        ops[0].tracking(cg),
+                        ops[1].tracking(cg),
+                    }),
```
```suggestion
                    error.SelectFailed => return cg.fail("failed to select {s} {} {} {}", .{
                        @tagName(tag),
                        cg.typeOf(bin_op.lhs).fmt(pt),
                        ops[0].tracking(cg),
                        ops[1].tracking(cg),
                    }),
```

jacobly0 commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                                .unused,
+                            },
+                            .each = .{ .once = &.{
+                                .{ ._, ._, .movzx, .dst0d, .src0b, ._, ._ },
```
Eliding zero extensions also applies here.  You probably don't even need any mutable inputs or temps, I'm guessing a non-aliasing dst0 is plenty.

bnuuydev replied:
Oh yes, I took a mental note of this but forgot to change it. A zero extension of the mask conditional on the instruction being pext is sufficient here. We could alias the destination reg with the mask here if the mask is in a register right?

jacobly0 replied:
Yep, being aliased just means that once you write to dst0, you can no longer read from srcn.

jacobly0 commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                                        else => unreachable,
+                                    } ++ "_" ++ "bigint",
+                                } } },
+                                .{ .type = .usize, .kind = .{ .reg = .rdi } },
+                                .{ .type = .usize, .kind = .{ .reg = .rsi } },
+                                .{ .type = .usize, .kind = .{ .reg = .rdx } },
+                                .{ .type = .usize, .kind = .{ .reg = .rcx } },
```
```suggestion
                                .{ .type = .usize, .kind = .{ .param_gpr = .{ .cc = .ccc, .index = 0 } } },
                                .{ .type = .usize, .kind = .{ .param_gpr = .{ .cc = .ccc, .index = 1 } } },
                                .{ .type = .usize, .kind = .{ .param_gpr = .{ .cc = .ccc, .index = 2 } } },
                                .{ .type = .usize, .kind = .{ .param_gpr = .{ .cc = .ccc, .index = 3 } } },
```

jacobly0 commented on lib/std/zig/llvm/Builder.zig:
```diff
@@ -4020,6 +4026,43 @@ pub const Intrinsic = enum {
             .attrs = &.{ .nocallback, .nofree, .nosync, .nounwind, .willreturn },
         },
+
+        .@"x86.bmi.pext.32" = .{
+            .ret_len = 1,
+            .params = &.{
+                .{ .kind = .{ .type = .i32 } },
+                .{ .kind = .{ .type = .i32 } },
+                .{ .kind = .{ .type = .i32 } },
+            },
+            .attrs = &.{ .nocallback, .nofree, .nosync, .nounwind, .{ .memory = Attribute.Memory.all(.none) } },
+        },
```
```suggestion
        .@"x86.bmi.pext = .{
            .ret_len = 1,
            .params = &.{
                .{ .kind = .{ .matches_scalar = 2 } },
                .{ .kind = .{ .matches_scalar = 2 } },
                .{ .kind = .overloaded },
            },
            .attrs = &.{ .nocallback, .nofree, .nosync, .nounwind, .willreturn, .{ .memory = .all(.none) } },
        },
```
etc.

jacobly0 replied:
Oh right, these don't follow the correct pattern...

bnuuydev replied:
Besides the other changes here, aren't you right to add `.willreturn` though?

jacobly0 reviewed (changes requested):
There's probably more issues, but they will be easier to see with these more obvious things fixed first.

bnuuydev commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                            .required_features = .{ .bmi2, null, null, null },
+                            .src_constraints = .{ .{ .int = .dword }, .{ .int = .dword }, .any },
+                            .patterns = &.{
+                                .{ .src = .{ .to_mut_gpr, .mem, .none } },
+                                .{ .src = .{ .mem, .to_mut_gpr, .none }, .commute = .{ 0, 1 } },
+                                .{ .src = .{ .to_mut_gpr, .to_gpr, .none } },
```
Oh, this is the part of the commit that wasn't written by me. I assume just `.to_gpr` is sufficient?

jacobly0 replied:
> `VEX.LZ.F2.0F38.W0 F5 /r PDEP r32a, r32b, r/m32`

This says that the second operand (src0) must be a register.

bnuuydev replied:
Oh yeah, I see, so `.{ .to_gpr, .mem, .none }` and `.{ .to_gpr, .to_gpr, .none }` are the only two patterns.

jacobly0 replied:
Yeah, `commute` means that it's valid to swap the source with the mask and get the same result!

bnuuydev commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                                .{ .src = .{ .mem, .to_mut_gpr, .none }, .commute = .{ 0, 1 } },
+                                .{ .src = .{ .to_mut_gpr, .to_gpr, .none } },
+                            },
+                            .dst_temps = .{ .{ .ref = .src0 }, .unused },
```
Same as above. We just want `.gpr` right? But also `pext` *can* alias registers, but that is handled by the register allocator right?

jacobly0 replied:
`.{ .rc = .general_purpose }` allocates a non-aliasing register, used when it must not alias any inputs
`.{ .mut_rc = .{ .rc = .general_purpose, .ref = .src0 } }` allocates a maybe-aliasing register, used when it is legal to alias src0, but src0 might pattern match to either a non-register or a non-mutable register
`.{ .ref = .src0 }` allocates a definitely-aliasing register, used when the pattern already forces a mutable register

bnuuydev replied:
Oh, a quick point of clarification. When we use `.mut_rc`, must the source register be `.mut_gpr`, or is `.gpr` sufficient?

jacobly0 replied:
You can ref `.gpr` from `.mut_rc`, and then if the referenced input dies on this instruction it will reuse that register, otherwise it will pick a new register.

bnuuydev replied:
Is there a way to say that it could alias either `.src0` or `.src1`?

bnuuydev replied:
For the `dword` and `qword` invocations of the `pext` and `pdep` instructions, there's no reason why the destination register couldn't alias either one of the source registers. Is there a way to encode that?

jacobly0 replied:
Just allow aliasing with the lowest numbered src that is correct.

bnuuydev commented on src/arch/x86_64/CodeGen.zig:
```diff
@@ -85897,6 +85897,339 @@ fn genBody(cg: *CodeGen, body: []const Air.Inst.Index) InnerError!void {
+                const bin_op = air_datas[@intFromEnum(inst)].bin_op;
+                var ops = try cg.tempsFromOperands(inst, .{ bin_op.lhs, bin_op.rhs });
+                var res: [1]Temp = undefined;
+                cg.select(&res, &.{cg.typeOf(bin_op.lhs)}, &ops, switch (tag) {
+                    inline .deposit_bits, .extract_bits => |air_tag| comptime &.{
```
Looks good! I'll get this and others written up in an hour or so.

bnuuydev commented:
I think I've addressed those all now :) I'll take a look at any feedback you've given tomorrow. Thank you for the help so-far! Also it'd probably be a good idea to give the workflows approval in-case any tests aren't passing on platforms I haven't been able to test for :)

bnuuydev commented:
Fixed the faulty assumption that the source and mask would have the same number of limbs in the `big.int` implementation. Also added a couple more tests to confirm we have the correct behaviour.

alexrp commented on lib/compiler_rt/pdeppext.zig:
```diff
@@ -0,0 +1,312 @@
+    return pext_uX(u128, source, mask);
+}
+
+// BEGIN HACKY CODE COPY WAIT FOR ALEXRP PR
```
#23483

bnuuydev commented on lib/std/math/big/int.zig:
```diff
@@ -1760,6 +1760,73 @@ pub const Mutable = struct {
+        assert(mask.positive);
+
+        result.positive = true;
+        @memset(result.limbs, 0);
```
May you clarify this please?

jacobly0 replied:
You could pass in a `result` with a billion limbs, and the `@memset` would require each one to be set to 0 and then the `normalize` would have to iterate each one to make sure they are still null.  Instead, you should only access the prefix of the limbs that might possibly be nonzero, which certainly does not exceed the number of limbs in the mask, among other constraints.

bnuuydev replied:
I see, I guess we could memset as we iterate the limbs instead then, so we only memset the limbs we actually write to? Or we could memset based on the length if the source or mask? Not sure which approach would be preferable here, but I like the first?

bnuuydev replied:
In fact for depositBits a memset isn't necessary at all  we can just use = instead of |= on the limb.

jacobly0 replied:
Both impls only ever access the limbs monotonically, so you just store the current limb in a local variable, and then when you need to move on to the next limb, you store the local into the next element and track how many initialized elements there are, and then the `normalize` only ever has to check the actually initialized limbs for being zero.

bnuuydev commented on lib/std/math/big/int_test.zig:
```diff
@@ -2960,6 +2960,59 @@ fn popCountTest(val: *const Managed, bit_count: usize, expected: usize) !void {
     try testing.expectEqual(expected, val.toConst().popCount(bit_count));
 }
 
+test "big int extractBits" {
```
Something like the gcd and bitwise or tests?

jacobly0 replied:
There's definitely room for improvement here relative to the existing tests, for example an `isNormalized` function and an expect that the return of any function is `true`, but anything that catches possible bugs is good enough for now.

bnuuydev replied:
I'm still a little bit stuck here with working out what you're actually asking for here, but I'll have another look at it soon. For instance, they're always going to return a normalised output, since they call `normalize` before returning, no?

bnuuydev commented:
Currently away from my computer, will get these changes addressed in a week or two :)

bnuuydev commented on lib/zig.h:
```diff
@@ -1381,6 +1381,92 @@ zig_builtin_clz(16)
 
+#define zig_builtin_extract_bits(w) \
+    static inline uint##w##_t zig_extract_bits_u##w(uint##w##_t source, uint##w##_t mask_) { \
+        uint##w##_t bb = 1;\
```
Feel like `current_bit` makes sense here, as it's the current bit of the result that we are deciding whether to set.

lin72h commented:
Just a gentle ping — the 0.15 release is [scheduled for August](https://www.youtube.com/watch?v=x3hOiOcbgeA). Is there any chance this PR could make it in time?
